telosnetwork/telos-wallet
# adding fallback when indexer is down to fetch nfts owners

# Fixes #659

## Description
This Pull Request addresses the issue of stale NFT indexer data. It implements a backup query mechanism and user notifications for cases where the indexer data is stale. The main enhancements include:
- Adding an indexer health check.
- Implementing backup queries for fetching the current NFT owner.
- Notifying users on the collection inventory page about the indexer being not healthy.
- Querying ERC721 and ERC1155 contracts directly for `ownerOf` and `balanceOf` updates when needed.
These changes ensure that NFT ownership and quantity are accurately reflected, even when the indexer data is not up-to-date.

## Test Scenarios
Currently, the indexer is up to date so I created a way to simulate the indexer is not healthy.
- go to [this page](https://deploy-preview-671--wallet-staging.netlify.app/) and login
- go to demo section (left-bottom) 
- go to "Simulation indexer down"
- press the button "Simulate indexer down" and wait for it to go down
- press Homepage link (don't refresh the page)
- login again
- go to inventory page
  - you should get a warning notification about indexer being down
- go to any ERC1155 detail page
- check the owner list
- transfer the NFT and wait for the page to reload NFT owners
  - it only updates your balance of that ERC1155
  - other owner balances stay the same (because we don't use the indexer)

## Videocapture

https://github.com/telosnetwork/telos-wallet/assets/4420760/5a742f1e-8daa-49fa-9a66-a68208814396



## Checklist:
-   [x] Implemented indexer health check and user notifications for stale data.
-   [x] Added backup queries to fetch current NFT data from contracts directly.
-   [x] Ensured that the solution works for both ERC721 and ERC1155 contracts.
-   [x] Conducted thorough testing to confirm the accuracy of NFT data.
-   [x] Reviewed and cleaned up the code in the affected areas.
-   [x] No new warnings generated by the changes.
-   [x] All changes merged and published in downstream modules.
-   [x] Code and comments reviewed for clarity and comprehensibility.
-   [x] English text included in the translation file or new issue created for required translations.
-   [x] Mobile functionality and responsiveness tested.
-   [x] Appropriate test coverage added.
